### PR TITLE
Export sslHandle in `net` and `asyncnet`.

### LINKED ISSUE #17021: Export sslHandle in net and asyncnet
### Summary

Export the `sslHandle` field of sockets in `net` and `asyncnet` respectively,
just like `SslContext` exports `context`.

### Description
Currently, the only way of using libssl procedures that take an ssl handle as a parameter(be them from `openssl` or imported through FFI separately) is by either rewritting the socket wrapping procedure or `include`-ing `net` or `asyncnet`, both of which being quite overkill for achieving simple tasks, even calling simple procedures from the stdlib `openssl` such as [SSL_get_error](https://nim-lang.org/docs/openssl.html#SSL_get_error%2CSslPtr%2Ccint) or [SSL_get_peer_certificate](https://nim-lang.org/docs/openssl.html#SSL_get_peer_certificate%2CSslCtx).

The `SslContext` itself allows us to access the low-level `context`, why shouldn't `Socket`?

### Alternatives

One could add the needed procedures to the nim openssl, but then comes the problem if it is actually needed, or if it keeps
compatibility with the rest, making development of both `openssl` and the program itself harder.

### Additional Information

This problem can be seen discussed [here](https://forum.nim-lang.org/t/7484)

### PR DESCRIPTION
**Pull request by ardek66**
Fix https://github.com/nim-lang/Nim/issues/17021

### CODE REVIEW

**Review by timotheecour — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Review by dom96 — CHANGES_REQUESTED** [review]
Exporting means that we allow read/write, we shouldn't.

Turn this into an accessor (`getSslHandle`).

For reviewers: please don't merge such changes. Let's consider carefully what we want to expose.

**Comment by konsumlamm** [comment]
> Turn this into an accessor (`getSslHandle`).

Why not just `sslHandle`?

File: lib/pure/asyncnet.nim [hunk 1]
```diff
@@ -731,6 +731,11 @@ proc close*(socket: AsyncSocket) =
         raiseSSLError()
 
 when defineSsl:
+  proc sslHandle*(socket: AsyncSocket): SslPtr =
+    ## Retrieve the ssl pointer of `socket`.
+    ## Useful for interfacing with `openssl`.
+    return socket.sslHandle
```
Comment (ghost) [inline lib/pure/asyncnet.nim]: Nit: return is redundant :)

Comment (timotheecour) [inline lib/pure/asyncnet.nim]:
```suggestion
  proc sslHandle*(self: AsyncSocket): SslPtr =
    ## Retrieve the ssl pointer of `self`.
    ## Useful for interfacing with `openssl`.
    self.sslHandle
```

(see nep1.rst)

File: lib/pure/net.nim [hunk 2]
```diff
@@ -501,6 +501,11 @@ when defineSsl:
   ERR_load_BIO_strings()
   OpenSSL_add_all_algorithms()
 
+  proc sslHandle*(socket: Socket): SslPtr =
+    ## Retrieve the ssl pointer of `socket`.
+    ## Useful for interfacing with `openssl`.
+    return socket.sslHandle
```
Comment (timotheecour) [inline lib/pure/net.nim]:
```suggestion
  proc sslHandle*(self: Socket): SslPtr =
    ## Retrieve the ssl pointer of `self`.
    ## Useful for interfacing with `openssl`.
    self.sslHandle
```

**Review by timotheecour — APPROVED** [review]

**Review by dom96 — CHANGES_REQUESTED** [review]
Why not just `sslHandle`? Well, there is a good reason for this: you avoid the risk of having an infinite loop because you call the proc again by accident instead of accessing the field. I have had this happen and I think this is good enough reason to adopt this as a convention and include the "get" prefix.

**Comment by timotheecour** [comment]
> you avoid the risk of having an infinite loop because you call the proc again by accident instead of accessing the field

do you have a reproducible example?
this works:
```nim
type Foo = object
  bar: int
proc bar(a: Foo): int = a.bar
let a = Foo(bar: 3)
echo a.bar
echo bar(a)
```

**Comment by timotheecour** [comment]
I'm adding a guideline in https://github.com/nim-lang/Nim/pull/17645 to discuss this, please comment in that PR

**Comment by ardek66** [comment]
@dom96

**Comment by dom96** [comment]
I don't recall the exact circumstances when this problem occurred, but other things to try are generics and defining these procs across module boundaries. I definitely had an infinite loop happen in the past which made me wary of naming the proc with the same name as the field, even if it was a bug in the past that has been fixed (or I dreamed this) it may be worthwhile to be defensive here (as reading the code itself can be confusing).

**Comment by ardek66** [comment]
@dom96 I tested it and it seems fields are checked before function calls.
https://play.nim-lang.org/#ix=3lHH
You can see that the function is only called at the second assignment of `bar`, when we explicitly write the call syntax.
I also tested it across modules, making a module with the type and proc definititions, and a program importing that and doing the calls.

```nim
# a.nim
type Foo* = object
  a*: int

proc a*(self: Foo): int =
  echo "the proc is called"
  return self.a
```
```nim
# b.nim
import a

var foo: Foo
foo.a = 3
var bar = foo.a
echo bar

bar = a(foo)
echo bar
```

Works exactly like the playground I presented above.
Knowing that fields are called before procs, I dont see why the ssl getter would lead into an infinite recursion error.

I think your infinite recursion was caused by the getter being defined in a context where you had a module in which an object was exported, but not the field the getter was supposed to wrap. Something like:
```nim
# a.nim

type Foo* = object
  a: int # Notice we didn't export this field
```

```nim
# b.nim
import a

var foo: Foo

proc a(foo: Foo): int =
  return foo.a # Since the field wasnt actually exported, this actually calls the procedure, resulting in an infinite loop

var bar = foo.a
echo bar

bar = a(foo)
echo bar

**Comment by ardek66** [comment]
To add to this even the manual states that the object field is preffered before the function call, so it shouldn't do recursion.
https://nim-lang.org/docs/manual.html#procedures-properties

**Review by dom96 — APPROVED** [review]
Since others are happy, feel free to merge :)

**Pull request merged**